### ISSUE TITLE
[Bug]: Output of a dynamic item with an existing key is silently ignored

### ISSUE THREAD

**Issue by asumagic** [speechbrain/speechbrain#1938]
### Describe the bug

Using `DynamicItemDataset`, when you add a dynamic item that provides a key that is already in the dataset, then the associated function _will_ be executed, but the output value will be ignored when you query the dataset for that key.

### Expected behaviour

It would probably be best for that situation to generate an error, as having a single key existing multiple times within the pipeline is probably prone to errors even in the case where it might be intended (e.g. to transform a value with an identical semantic meaning through the pipeline).

### To Reproduce

```python
from speechbrain.dataio.dataset import DynamicItemDataset

test_dd = DynamicItemDataset(data={
    "test_key": {
        "foo": "a"
    }
})

def dummy_func(_foo):
    print(f"dummy_func invoked with {_foo}")
    return "b"

test_dd.add_dynamic_item(dummy_func, takes=["foo"], provides="foo")
# test_dd.add_dynamic_item(lambda _foo: "b", takes=["foo"], provides="bar")
test_dd.set_output_keys(["foo"])

print(list(iter(test_dd)))
```

The output is:

```
dummy_func invoked with a
[{'foo': 'a'}]
```

The function was invoked, but the original value for key `"foo"` was retained.

### Versions

Tip of `develop` `e353e95` is affected.

**Comment by Gastron** [comment]
Agreed, adding an item with a key that already exists should simply raise an error.

Unfortunately this pattern is used in many recipes currently. Should we make a breaking change and try to change any recipes we find that use this?

**Comment by Adel-Moumen** [comment]
Hi @AsuMagic what do you think about what Gastron said?
